File: core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/managers/StatisticsManager.scala
package edu.uci.ics.amber.engine.architecture.worker.managers

import edu.uci.ics.amber.core.executor.{OperatorExecutor, SinkOperatorExecutor}
import edu.uci.ics.amber.engine.architecture.worker.statistics.{
  PortTupleMetricsMapping,
  TupleMetrics,
  WorkerStatistics
}
import edu.uci.ics.amber.core.workflow.PortIdentity

import scala.collection.mutable

class StatisticsManager {
  // DataProcessor
  private val inputStatistics: mutable.Map[PortIdentity, (Long, Long)] =
    mutable.Map.empty.withDefaultValue((0L, 0L))
  private val outputStatistics: mutable.Map[PortIdentity, (Long, Long)] =
    mutable.Map.empty.withDefaultValue((0L, 0L))
  private var dataProcessingTime: Long = 0L
  private var totalExecutionTime: Long = 0L
  private var workerStartTime: Long = 0L

  // AmberProcessor
  private var controlProcessingTime: Long = 0L

  /**
    * Retrieves the current statistics for the operator.
    * @param operator the operator executor
    * @return a WorkerStatistics object containing the statistics
    */
  def getStatistics(operator: OperatorExecutor): WorkerStatistics = {
    val userFriendlyOutputStatistics = operator match {
      case _: SinkOperatorExecutor => inputStatistics
      case _                       => outputStatistics
    }

    WorkerStatistics(
      inputStatistics.map {
        case (portId, (tupleCount, tupleSize)) =>
          PortTupleMetricsMapping(portId, TupleMetrics(tupleCount, tupleSize))
      }.toSeq,
      userFriendlyOutputStatistics.map {
        case (portId, (tupleCount, tupleSize)) =>
          PortTupleMetricsMapping(portId, TupleMetrics(tupleCount, tupleSize))
      }.toSeq,
      dataProcessingTime,
      controlProcessingTime,
      totalExecutionTime - dataProcessingTime - controlProcessingTime
    )
  }

  /**
    * Calculates the total number of input tuples.
    * @return the total input tuple count
    */
  def getInputTupleCount: Long = inputStatistics.values.map(_._1).sum

  /**
    * Calculates the total number of output tuples.
    * @return the total output tuple count
    */
  def getOutputTupleCount: Long = outputStatistics.values.map(_._1).sum

  /**
    * Increases the input statistics for a given port.
    * @param portId the port identity
    * @param size the size of the tuple
    */
  def increaseInputStatistics(portId: PortIdentity, size: Long): Unit = {
    require(size >= 0, "Tuple size must be non-negative")
    val (count, totalSize) = inputStatistics(portId)
    inputStatistics.update(portId, (count + 1, totalSize + size))
  }

  /**
    * Increases the output statistics for a given port.
    * @param portId the port identity
    * @param size the size of the tuple
    */
  def increaseOutputStatistics(portId: PortIdentity, size: Long): Unit = {
    require(size >= 0, "Tuple size must be non-negative")
    val (count, totalSize) = outputStatistics(portId)
    outputStatistics.update(portId, (count + 1, totalSize + size))
  }

  /**
    * Increases the data processing time.
    * @param time the time to add
    */
  def increaseDataProcessingTime(time: Long): Unit = {
    require(time >= 0, "Time must be non-negative")
    dataProcessingTime += time
  }

  /**
    * Increases the control processing time.
    * @param time the time to add
    */
  def increaseControlProcessingTime(time: Long): Unit = {
    require(time >= 0, "Time must be non-negative")
    controlProcessingTime += time
  }

  /**
    * Updates the total execution time.
    * @param time the current time
    */
  def updateTotalExecutionTime(time: Long): Unit = {
    require(
      time >= workerStartTime,
      "Current time must be greater than or equal to worker start time"
    )
    totalExecutionTime = time - workerStartTime
  }

  /**
    * Initializes the worker start time.
    * @param time the start time
    */
  def initializeWorkerStartTime(time: Long): Unit = {
    workerStartTime = time
  }
}

File: amber/src/main/python/core/architecture/managers/statistics_manager.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from collections import defaultdict
from typing import DefaultDict

from loguru import logger

from proto.org.apache.texera.amber.core import PortIdentity
from proto.org.apache.texera.amber.engine.architecture.worker import (
    WorkerStatistics,
    PortTupleMetricsMapping,
    TupleMetrics,
)


class StatisticsManager:
    def __init__(self) -> None:
        # Initialize metrics with default values
        self._input_tuple_metrics: DefaultDict[PortIdentity, TupleMetrics] = (
            defaultdict(lambda: (0, 0))
        )
        self._output_tuple_metrics: DefaultDict[PortIdentity, TupleMetrics] = (
            defaultdict(lambda: (0, 0))
        )
        self._data_processing_time: int = 0
        self._control_processing_time: int = 0
        self._total_execution_time: int = 0
        self._worker_start_time: int = 0

    def get_statistics(self) -> WorkerStatistics:
        # Compile and return worker statistics
        return WorkerStatistics(
            [
                PortTupleMetricsMapping(port_id, TupleMetrics(*tuple_metrics))
                for port_id, tuple_metrics in self._input_tuple_metrics.items()
            ],
            [
                PortTupleMetricsMapping(port_id, TupleMetrics(*tuple_metrics))
                for port_id, tuple_metrics in self._output_tuple_metrics.items()
            ],
            self._data_processing_time,
            self._control_processing_time,
            max(
                0,
                self._total_execution_time
                - self._data_processing_time
                - self._control_processing_time,
            ),
        )

    def increase_input_statistics(self, port_id: PortIdentity, size: int) -> None:
        if size < 0:
            raise ValueError("Tuple size must be non-negative")
        count, total_size = self._input_tuple_metrics[port_id]
        self._input_tuple_metrics[port_id] = (count + 1, total_size + size)

    def increase_output_statistics(self, port_id: PortIdentity, size: int) -> None:
        if size < 0:
            raise ValueError("Tuple size must be non-negative")
        count, total_size = self._output_tuple_metrics[port_id]
        self._output_tuple_metrics[port_id] = (count + 1, total_size + size)

    def increase_data_processing_time(self, time: int) -> None:
        if time < 0:
            raise ValueError("Time must be non-negative")
        self._data_processing_time += time

    def increase_control_processing_time(self, time: int) -> None:
        if time < 0:
            raise ValueError("Time must be non-negative")
        self._control_processing_time += time

    def update_total_execution_time(self, time: int) -> None:
        if time < self._worker_start_time:
            raise ValueError(
                "Current time must be greater than or equal to worker start time"
            )
        new_total = time - self._worker_start_time
        if new_total < self._total_execution_time:
            logger.warning(
                f"update_total_execution_time called with non-monotonic time: "
                f"new total {new_total}ns < current total {self._total_execution_time}ns. "
                "Clock skew or out-of-order call detected."
            )
        processing_total = self._data_processing_time + self._control_processing_time
        if new_total < processing_total:
            logger.warning(
                f"idle_time drift: total_execution_time ({new_total}ns) < "
                f"data ({self._data_processing_time}ns) + control "
                f"({self._control_processing_time}ns). "
                "update_total_execution_time should be called after increase_*_processing_time "
                "with the same end timestamp. idle_time will be clamped to 0."
            )
        self._total_execution_time = new_total

    def initialize_worker_start_time(self, time: int) -> None:
        # Set the worker start time
        self._worker_start_time = time
